pkg/jobstore/boltdb/store.go

```diff
@@ -0,0 +1,1013 @@
+		return err
+	} else {
+		seq := BucketSequenceString(tx, bkt)
+		if err = bkt.Put([]byte(seq), data); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// CreateExecution creates a record of a new execution
+func (b *BoltJobStore) CreateExecution(ctx context.Context, execution model.ExecutionState) error {
+	return b.database.Update(func(tx *bolt.Tx) (err error) {
+		return b.createExecution(tx, execution)
+	})
+}
+
+func (b *BoltJobStore) createExecution(tx *bolt.Tx, execution model.ExecutionState) error {
+	if !b.jobExists(tx, execution.JobID) {
+		return jobstore.NewErrJobNotFound(execution.JobID)
+	}
+
+	execID := []byte(execution.ID().String())
+
+	if execution.CreateTime.IsZero() {
+		execution.CreateTime = b.clock.Now().UTC()
+	}
+	if execution.UpdateTime.IsZero() {
+		execution.UpdateTime = execution.CreateTime
+	}
+	if execution.Version == 0 {
+		execution.Version = 1
+	}
+
+	// Check for the existence of this ID and if it doesn't already exist, then create
+	// it
+	if bucket, err := NewBucketPath(BucketPathExecutions).Get(tx, false); err != nil {
+		return err
+	} else {
+		_, err := b.getExecution(tx, execution.ID())
+		if err == nil {
+			return jobstore.NewErrExecutionAlreadyExists(execution.ID())
+		}
+
+		if data, err := json.Marshal(execution); err != nil {
+			return err
+		} else {
+			err = bucket.Put(execID, data)
+			if err != nil {
+				return err
+			}
+
+			b.triggerEvent(jobstore.ExecutionWatcher, jobstore.CreateEvent, data)
+		}
+	}
+
+	return b.appendExecutionHistory(tx, execution, model.ExecutionStateNew, "")
+}
+
+// UpdateExecution updates the state of a single execution by loading from storage,
+// updating and then writing back in a single transaction
+func (b *BoltJobStore) UpdateExecution(ctx context.Context, request jobstore.UpdateExecutionRequest) error {
+	return b.database.Update(func(tx *bolt.Tx) (err error) {
+		return b.updateExecution(tx, request)
+	})
+}
+
+func (b *BoltJobStore) updateExecution(tx *bolt.Tx, request jobstore.UpdateExecutionRequest) error {
+	if !b.jobExists(tx, request.ExecutionID.JobID) {
+		return jobstore.NewErrJobNotFound(request.ExecutionID.JobID)
+	}
+
+	existingExecution, err := b.getExecution(tx, request.ExecutionID)
+	if err != nil {
+		return jobstore.NewErrExecutionNotFound(request.ExecutionID)
+	}
+
+	// check the expected state
+	if err := request.Condition.Validate(existingExecution); err != nil {
+		return err
+	}
+	if existingExecution.State.IsTerminal() {
+		return jobstore.NewErrExecutionAlreadyTerminal(request.ExecutionID, existingExecution.State, request.NewValues.State)
+	}
+
+	// populate default values
+	newExecution := request.NewValues
+	if newExecution.CreateTime.IsZero() {
+		newExecution.CreateTime = b.clock.Now().UTC()
+	}
+	if newExecution.UpdateTime.IsZero() {
+		newExecution.UpdateTime = existingExecution.CreateTime
+	}
+	if newExecution.Version == 0 {
+		newExecution.Version = existingExecution.Version + 1
+	}
```

Comment:
we should use createTime of the existing execution, and now() for the update time as the default values.

Reply:
Fixed in the InMemory store as well.